src/docmanager/index.ts

```diff
@@ -0,0 +1,958 @@
+      return;
+    }
+    let lang = mFactoryEx.factory.preferredLanguage(path);
+    let model = mFactoryEx.factory.createNew(lang);
+    let opts = utils.copy(mFactoryEx.contentsOptions);
+    if (opts.format === 'json') {
+      opts.content = model.toJSON();
+    } else {
+      opts.content = model.toString();
+    }
+    manager.save(path, opts).then(contents => {
+      let id = this._createContext(path, model, widgetName, contents);
+      this._createWidget(id, widgetName, widget, kernel);
+    });
+    return widget;
+  }
+
+  /**
+   * Get the path given a widget.
+   */
+  getPath(widget: Widget): string {
+    let id = Private.contextProperty.get(widget);
+    return this._contextManager.getPath(id);
+  }
+
+  /**
+   * Clone a widget.
+   *
+   * #### Notes
+   * This will create a new widget with the same model and context
+   * as the existing widget.
+   */
+  clone(widget: Widget): Widget {
+    let parent = new Widget();
+    let id = Private.contextProperty.get(widget);
+    let name = Private.nameProperty.get(widget);
+    this._createWidget(id, name, parent);
+    return parent;
+  }
+
+  /**
+   * Filter messages on the widget.
+   */
+  filterMessage(handler: IMessageHandler, msg: Message): boolean {
+    if (msg.type !== 'close-request') {
+      return false;
+    }
+    if (this._closeGuard) {
+      // Allow the close to propagate to the widget and its layout.
+      this._closeGuard = false;
+      return false;
+    }
+    let widget = handler as Widget;
+    let id = Private.contextProperty.get(widget);
+    let model = this._contextManager.getModel(id);
+    let context = this._contextManager.getContext(id);
+    let child = (widget.layout as PanelLayout).childAt(0);
+    let name = Private.nameProperty.get(widget);
+    // Check for a sibling widget.
+    if (!name) {
+      // Do not filter the message.
+      return false;
+    }
+    let factory = this._widgetFactories[name].factory;
+    this._maybeClose(widget, model.dirty).then(result => {
+      if (!result) {
+        return result;
+      }
+      return factory.beforeClose(model, context, child);
+    }).then(result => {
+      if (result) {
+        return this._cleanupWidget(widget);
+      }
+    });
+    return true;
+  }
+
+  /**
+   * Update the path of an open document.
+   *
+   * @param oldPath - The previous path.
+   *
+   * @param newPath - The new path.
+   */
+  renameFile(oldPath: string, newPath: string): void {
+    let ids = this._contextManager.getIdsForPath(oldPath);
+    for (let id in ids) {
+      this._contextManager.rename(id, newPath);
+    }
+  }
+
+  /**
+   * Handle a file deletion on the currently open widgets.
+   *
+   * @param path - The path of the file to delete.
+   */
+  deleteFile(path: string): void {
+    let ids = this._contextManager.getIdsForPath(path);
+    for (let id of ids) {
+      let widgets: Widget[] = this._widgets[id] || [];
+      for (let w of widgets) {
+        this._cleanupWidget(w);
```

Comment:
I think it would be much simpler and more efficient to just loop through the widgets, call `widget.dispose()` on each widget (which should call `widget.close` automatically, right?), then dispose of each context. At what point are the models disposed?